diff -cr a/_version.py b/_version.py
*** a/_version.py	Mon Aug 28 15:22:04 2017
--- b/_version.py	Mon Aug 28 15:22:43 2017
***************
*** 28,33 ****
--- 28,34 ----
          from distutils.version import LooseVersion as V
          hub_major_minor = V(hub_version).version[:2]
          singleuser_major_minor = V(singleuser_version).version[:2]
+         extra = ""
          if singleuser_major_minor == hub_major_minor:
              # patch-level mismatch or lower, log difference at debug-level
              # because this should be fine
