diff -cr a/eups b/eups
*** a/eups	Tue Sep  5 10:24:25 2017
--- b/eups	Tue Sep  5 10:30:34 2017
***************
*** 12,20 ****
      import eups.cmd
  except ImportError:
      eupsdir = None
!     if os.environ.has_key("EUPS_DIR"):
!         eupsdir = os.environ["EUPS_DIR"]
!     else:
          # the first item on sys.path is the script directory (bin)
          eupsdir = os.path.dirname(sys.path[0])
          if not os.path.isabs(eupsdir):
--- 12,19 ----
      import eups.cmd
  except ImportError:
      eupsdir = None
!     eupsdir = os.getenv("EUPS_DIR")
!     if not eupsdir:
          # the first item on sys.path is the script directory (bin)
          eupsdir = os.path.dirname(sys.path[0])
          if not os.path.isabs(eupsdir):
