diff -cr a/handlers.py b/handlers.py
*** a/handlers.py	Mon Aug 21 13:51:45 2017
--- b/handlers.py	Mon Aug 21 13:52:46 2017
***************
*** 83,89 ****
          page_config.setdefault(
              'assetsDir', config.assets_dir.replace(os.sep, '/')
          )
! 
          if os.path.exists(page_config_file):
              with open(page_config_file) as fid:
                  try:
--- 83,98 ----
          page_config.setdefault(
              'assetsDir', config.assets_dir.replace(os.sep, '/')
          )
!         # Grab API token from environment
!         api_token = os.getenv('JUPYTERLAB_API_TOKEN')
!         if not api_token:
!             api_token = ''  # Turn None into empty string
!         page_config.setdefault('token', api_token)
!         if not self.token:
!             try:
!                 self.token = api_token
!             except AttributeError:
!                 self.log.error("Can't set self.token.")
          if os.path.exists(page_config_file):
              with open(page_config_file) as fid:
                  try:
