diff -cr a/labhubapp.py b/labhubapp.py
*** a/labhubapp.py	Fri Jul 14 14:11:40 2017
--- b/labhubapp.py	Fri Jul 14 14:12:04 2017
***************
*** 1,3 ****
--- 1,4 ----
+ import os
  from .labapp import LabApp
  
  try:
***************
*** 7,12 ****
--- 8,14 ----
      raise ImportError('You must have jupyterhub installed for this to work.')
  else:
      class SingleUserLabApp(SingleUserNotebookApp, LabApp):
+ 
          def init_webapp(self, *args, **kwargs):
              super().init_webapp(*args, **kwargs)
              settings = self.web_app.settings
***************
*** 15,20 ****
--- 17,31 ----
              settings['page_config_data']['hub_prefix'] = self.hub_prefix
              settings['page_config_data']['hub_host'] = self.hub_host
              settings['page_config_data']['hub_user'] = self.user
+             api_token = os.getenv('JUPYTERHUB_API_TOKEN')
+             if not api_token:
+                 api_token = ''
+             if not self.token:
+                 try:
+                     self.token = api_token
+                 except AttributeError:
+                     self.log.error("Can't set self.token")
+             settings['page_config_data']['token'] = api_token
  
  
  def main(argv=None):
