diff -cr a/base.py b/base.py
*** a/base.py	Fri Jul 21 14:01:18 2017
--- b/base.py	Fri Jul 21 14:06:46 2017
***************
*** 229,236 ****
          orm_user = orm.User.find(db=self.db, name=name)
          return self._user_from_orm(orm_user)
  
!     def user_from_username(self, username):
          """Get User for username, creating if it doesn't exist"""
          user = self.find_user(username)
          if user is None:
              # not found, create and register user
--- 229,240 ----
          orm_user = orm.User.find(db=self.db, name=name)
          return self._user_from_orm(orm_user)
  
!     def user_from_username(self, nameobj):
          """Get User for username, creating if it doesn't exist"""
+         if type(nameobj) is dict:
+             username = nameobj["name"]
+         else:
+             username = nameobj # This will explode if it's not a string
          user = self.find_user(username)
          if user is None:
              # not found, create and register user
