--- /usr/local/lib/python3.6/site-packages/kubernetes/config/dateutil.py	2019-12-20 10:42:03.000000000 -0700
+++ /tmp/dateutil.py	2019-12-20 10:40:55.000000000 -0700
@@ -58,7 +58,11 @@
     for x in range(6):
         dt[x] = int(groups[x])
     if groups[6] is not None:
-        dt[6] = int(groups[6])
+        group6 = groups[6]
+        while group6[0] == '.':
+            # Strip leading period from fractional second part
+            group6 = group6[1:]
+        dt[6] = int(group6)
